File: loja/produtos/src/main/java/br/com/bota/integration/ProdutoResource.java
package br.com.bota.integration;

import br.com.bota.dto.ProdutoDTO;
import br.com.bota.service.ProdutoService;
import io.micronaut.http.HttpResponse;
import io.micronaut.http.MediaType;
import io.micronaut.http.annotation.*;
import io.micronaut.tracing.annotation.ContinueSpan;
import io.micronaut.tracing.annotation.NewSpan;
import io.micronaut.tracing.annotation.SpanTag;

import javax.inject.Inject;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Controller("/produto")
public class ProdutoResource {
    @Inject
    private ProdutoService productService;

    @Get("/{id}")
    @ContinueSpan
    public HttpResponse<ProdutoDTO> findById(@SpanTag("produto.find.id") UUID id) {
        return HttpResponse.ok(ProdutoDTO.toDTO(productService.procurarPorId(id)));
    }

    @Get("/check/{id}")
    @ContinueSpan
    public HttpResponse<Boolean> checkQuantity(UUID id) {
        return HttpResponse.ok(productService.verificarQuantidade(id));
    }

    @Get
    @Produces(MediaType.APPLICATION_JSON)
    @ContinueSpan
    public HttpResponse<List<ProdutoDTO>> findAll() {
        var products = productService.procurar();
        return HttpResponse.ok(products.stream().map(ProdutoDTO::toDTO).collect(Collectors.toList()));
    }

    @Post
    @ContinueSpan
    public HttpResponse<ProdutoDTO> save(@SpanTag("produto.payload") ProdutoDTO dto) {
        var saved = productService.salvar(ProdutoDTO.toEntity(dto));
        return HttpResponse.created(ProdutoDTO.toDTO(saved));
    }

    @Patch
    @ContinueSpan
    public HttpResponse<ProdutoDTO> update(ProdutoDTO dto) {
        var saved = productService.salvar(ProdutoDTO.toEntity(dto));
        return HttpResponse.created(ProdutoDTO.toDTO(saved));
    }
}

File: loja-kotlin/kprodutos/src/main/kotlin/br/com/bota/resource/ProdutoResource.kt
package br.com.bota.resource

import br.com.bota.converter.ProdutoConverter
import br.com.bota.lojalib.dto.ProdutoDTO
import br.com.bota.service.ProdutoService
import br.com.bota.service.impl.ProdutoServiceImpl
import io.micronaut.http.HttpResponse
import io.micronaut.http.MediaType
import io.micronaut.http.annotation.*
import io.micronaut.tracing.annotation.ContinueSpan
import jakarta.inject.Inject
import java.util.*

interface ProdutoResource {
    fun findById(id: UUID): HttpResponse<ProdutoDTO>
    fun verificarQuantidade(id: UUID): HttpResponse<Boolean>
    fun findAll(): HttpResponse<List<ProdutoDTO>>
    fun save(dto: ProdutoDTO): HttpResponse<ProdutoDTO>
    fun update(dto: ProdutoDTO): HttpResponse<ProdutoDTO>
}

@Controller("/produto")
open class ProdutoResourceImpl @Inject constructor(private val produtoService: ProdutoService): ProdutoResource{
    @Get("/{id}")
    @ContinueSpan
    override fun findById(id: UUID): HttpResponse<ProdutoDTO> =
        HttpResponse.ok(ProdutoConverter.toProdutoDTO(produtoService.procurarPorId(id)))

    @Get("/check/{id}")
    @ContinueSpan
    override fun verificarQuantidade(id: UUID): HttpResponse<Boolean> =
        HttpResponse.ok(produtoService.verificarQuantidade(id))

    @Get
    @Produces(MediaType.APPLICATION_JSON)
    @ContinueSpan
    override fun findAll(): HttpResponse<List<ProdutoDTO>> {
        val produtos = produtoService.procurar()
        return HttpResponse.ok(produtos.map { ProdutoConverter.toProdutoDTO(it) })
    }

    @Post
    @ContinueSpan
    override fun save(dto: ProdutoDTO): HttpResponse<ProdutoDTO> =
        HttpResponse.created(ProdutoConverter.toProdutoDTO(produtoService.salvar(ProdutoConverter.toProduto(dto))))

    @Patch
    @ContinueSpan
    override fun update(dto: ProdutoDTO): HttpResponse<ProdutoDTO> {
        return HttpResponse.created(ProdutoConverter.toProdutoDTO(produtoService.salvar(ProdutoConverter.toProduto(dto))))
    }
}